text/0000-process-specific-env.md

```diff
@@ -0,0 +1,81 @@
+- CNB Issue: 
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Each process type could have its own independent environment variables.
+
+# Motivation
+[motivation]: #motivation
+
+- Greater flexibility to define runtime processes
+- When a new process type is created it often has different requirements for environment variables
+
+One example is a buildpack requiring different `NODE_OPTIONS` for a `web` process and a `debug` process. This would allow a node buildpack author to write `<layer>/env.launch/debug/NODE_OPTIONS` with a value of `--inspect=0.0.0.0:9229` for `debug` while omitting the file entirely for `web`. While this is currently possible for the buildpack author when defining the launch processes, this change would also allow ad-hoc `launcher` commands to ignore `envs` intended for specific processes.
+
+```
+# defined processes
+/cnb/lifecycle/launcher debug # NODE_OPTIONS=--inspect=0.0.0.0:9229
+/cnb/lifecycle/launcher web # no NODE_OPTIONS defined
+
+# ad-hoc
+/cnb/lifecycle/launcher -- npm install # no NODE_OPTIONS defined
+/cnb/lifecycle/launcher -- bash # no NODE_OPTIONS defined
+```
+
+`<layer>/profile.d/debug/debug.sh` would work similarly and be sourced by Bash before launching the `debug` command, but not when executing any other process.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Each env or profile layer dir can have subdirs like `<layer>/env.launch/web/` and `<layer>/profile.d/web/`. These behave the same way as `<layer>/env` but are specific to the launch of the `web` process.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Lifecycle would use the identifier of the intended launch process to bring in additional `env` and `profile.d` files. The launcher would sort these process specific values _after_ less specific values to allow [Environment Variable Modification Rules](https://github.com/buildpacks/spec/blob/master/buildpack.md#environment-variable-modification-rules) to apply as expected.
+
+The environment modification rules should be updated to reflect that sort order for `.prepend`, `.append`, and `.override` to reflect that process specific values are more processed in the most overriding order. The following shows an example of processing order for a given layer.
+
+```
+<layers>/<layer>/env/NODE_OPTIONS.append
+<layers>/<layer>/env.launch/NODE_OPTIONS.append
+<layers>/<layer>/env.launch/web/NODE_OPTIONS.append
+```
+
+Example spec update:
+
+- Environment variable file contents originating in the same layer MUST be sorted such that file contents in <layers>/<layer>/env.launch/<process>/ precede file contents in <layers>/<layer>/env.launch/, which must precede <layers>/<layer>/env/.
+- Environment variable file contents originating in the same layer MUST be sorted such that file contents in <layers>/<layer>/env/ precede file contents in <layers>/<layer>/env.build/ or <layers>/<layer>/env.launch/ which must preced file contents in <layers>/<layer>/env.launch/<process>/.

```

Comment:
```suggestion
- Environment variable file contents originating in the same layer MUST be sorted such that file contents in <layers>/<layer>/env/ precede file contents in <layers>/<layer>/env.build/ or <layers>/<layer>/env.launch/ which must preced file contents in <layers>/<layer>/env.launch/<process>/.
```suggestion
- Environment variable file contents originating in the same layer MUST be sorted such that file contents in `<layers>/<layer>/env/` precede file contents in `<layers>/<layer>/env.build/` or `<layers>/<layer>/env.launch/` which must precede file contents in `<layers>/<layer>/env.launch/<process>/`.

```
```